### ISSUE TITLE
Improve waiting for processes on Windows

### ISSUE THREAD

**Issue by ghost** [ocaml/dune#1289]
This is another small improvement that could help Windows builds. Currently, on Windows we resort to busy polling as `Unix` doesn't provide a way to wait for multiple processes:

```ocaml
  exception Finished of running_job * Unix.process_status

  let wait_nonblocking_win32 () =
    match
      Hashtbl.iter all ~f:(fun ~key:pid ~data:job ->
        let pid, status = Unix.waitpid [WNOHANG] pid in
        if pid <> 0 then
          raise_notrace (Finished (job, status)))
    with
    | () -> None
    | exception (Finished (job, status)) ->
      Hashtbl.remove all job.pid;
      Some (job, status)

  let rec wait_win32 () =
    match wait_nonblocking_win32 () with
    | None ->
      ignore (Unix.select [] [] [] 0.001);
      wait_win32 ()
    | Some x -> x
```

This could be avoided by using `WaitForMultipleObjects`.

**Comment by diml** [comment]
Yes. However, for every C stub we need an alternative OCaml implementation to use during bootstrap.

**Comment by rgrinberg** [comment]
ping @nojb. we're still polling and you always seem to run into performance problems with dune :)

**Comment by nojb** [comment]
@rgrinberg Thanks for the ping; I had forgotten about this issue. I'll try to give this one a try and report back.

**Comment by rgrinberg** [comment]
Np. In particular, it would be nice to have this as a feature in the spawn library. That way, I could make use of it in lsp and drop the polling there as well.

**Comment by nojb** [comment]
I have a first try at https://github.com/ocaml/dune/compare/main...nojb:WaitForMultipleObjects?expand=1. However in a quick test I didn't see any speedup (and it even seemed to be a tiny bit slower). Anyway I need to do more testing and make sure I didn't miss anything.

**Comment by rgrinberg** [comment]
Indeed your implementation seems fishy to me. When a new process is created, your call to `WaitForMultipleObjects` should be interrupted and done again with the new process list.

**Comment by nojb** [comment]
> Indeed your implementation seems fishy to me. When a new process is created, your call to `WaitForMultipleObjects` should be interrupted and done again with the new process list.

Thanks, sounds reasonable; I'll try to take another look soon.

**Comment by maiste** [comment]
@nojb I know this an old thread, but we are trying to go through the old issues. Is it still something you find relevant?

**Comment by nojb** [comment]
> [@nojb](https://github.com/nojb) I know this an old thread, but we are trying to go through the old issues. Is it still something you find relevant?

Not planning to work in the near term. Anyway, if one wanted to seriously improve performane on Windows, one would have to work instead towards trying to launch fewer processes for each compilation (eg by having long-lived compiler processes instead of launching the compiler once per compilation command).
